File: JDK11/opentraceability/src/main/java/opentraceability/models/events/eventbase.java
package opentraceability.models.events;

import opentraceability.Constants;
import opentraceability.interfaces.IEventKDE;
import opentraceability.models.events.kdes.CertificationList;
import opentraceability.utility.attributes.*;
import opentraceability.models.identifiers.PGLN;

import java.net.URI;
import java.time.*;

public class EventBase {
    @OpenTraceabilityAttribute(value = "", name = "eventTime", order = 1)
    public OffsetDateTime eventTime;

    @OpenTraceabilityAttribute(value = "", name = "recordTime", order = 2)
    public OffsetDateTime recordTime;

    @OpenTraceabilityAttribute(value = "", name = "eventTimeZoneOffset", order = 3)
    public Duration eventTimeZoneOffset;

    @OpenTraceabilityAttribute(value = "", name = "eventID", order = 4, versions = {EPCISVersion.V2, EPCISVersion.V1}, conditionalVersions = {})
    public URI eventID;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(value = "", name = "errorDeclaration", order = 5, versions = {EPCISVersion.V2, EPCISVersion.V1}, conditionalVersions = {})
    public ErrorDeclaration errorDeclaration;

    @OpenTraceabilityAttribute(value = "", name = "certificationInfo", order = 6, versions = {EPCISVersion.V2}, conditionalVersions = {})
    public String certificationInfo;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(value = Constants.CBVMDA_NAMESPACE, name = "certificationList", versions = {}, conditionalVersions = {})
    @OpenTraceabilityJsonAttribute("cbvmda:certificationList")
    public CertificationList certificationList;

    @OpenTraceabilityAttribute(value = Constants.CBVMDA_NAMESPACE, name = "informationProvider", versions = {}, conditionalVersions = {})
    @OpenTraceabilityJsonAttribute("cbvmda:informationProvider")
    public PGLN informationProvider;

    @OpenTraceabilityExtensionElementsAttribute
    public MutableList<IEventKDE> kdes = new MutableList<>();

    public <T extends IEventKDE> T getKDE(Class<T> clazz, String ns, String name) {
        IEventKDE kde = kdes.find(kde -> kde.namespace == ns && kde.name == name);
        if (kde != null) {
            if (clazz.isInstance(kde)) {
                @SuppressWarnings("unchecked")
                T result = (T) kde;
                return result;
            }
        }
        return null;
    }

    public <T extends IEventKDE> T getKDE(Class<T> clazz) {
        IEventKDE kde = kdes.find(kde -> clazz.isInstance(kde));
        if (kde != null) {
            @SuppressWarnings("unchecked")
            T result = (T) kde;
            return result;
        }
        return null;
    }
}

File: Java/OpenTraceability/src/main/opentraceability/models/events/eventbase.kt
package models.events

import interfaces.IEventKDE
import models.events.kdes.CertificationList
import Constants
import models.identifiers.PGLN
import utility.attributes.OpenTraceabilityAttribute
import utility.attributes.OpenTraceabilityExtensionElementsAttribute
import utility.attributes.OpenTraceabilityJsonAttribute
import utility.attributes.OpenTraceabilityObjectAttribute
import java.net.URI
import java.time.Duration
import java.time.OffsetDateTime

open class EventBase {
    @OpenTraceabilityAttribute("","eventTime", 1)
    var eventTime: OffsetDateTime? = null

    @OpenTraceabilityAttribute("","recordTime", 2)
    var recordTime: OffsetDateTime? = null

    @OpenTraceabilityAttribute("","eventTimeZoneOffset", 3)
    var eventTimeZoneOffset: Duration? = null

    @OpenTraceabilityAttribute("","eventID", 4, EPCISVersion.V2)
    //@OpenTraceabilityAttribute("","baseExtension/eventID", 4, EPCISVersion.V1)
    var eventID: URI? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("","errorDeclaration", 5, EPCISVersion.V2)
    //@OpenTraceabilityAttribute("","baseExtension/errorDeclaration", 5, EPCISVersion.V1)
    var errorDeclaration: ErrorDeclaration? = null

    @OpenTraceabilityAttribute("","certificationInfo", 6, EPCISVersion.V2)
    var certificationInfo: String? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(Constants.CBVMDA_NAMESPACE, "certificationList")
    @OpenTraceabilityJsonAttribute("cbvmda:certificationList")
    var certificationList: CertificationList? = null

    @OpenTraceabilityAttribute(Constants.CBVMDA_NAMESPACE, "informationProvider")
    @OpenTraceabilityJsonAttribute("cbvmda:informationProvider")
    var informationProvider: PGLN? = null

    @OpenTraceabilityExtensionElementsAttribute
    var kdes: MutableList<IEventKDE> = mutableListOf()

    inline fun <reified T: IEventKDE> getKDE(ns: String, name: String): T? {
        val kde = kdes.find { it.Namespace == ns && it.Name == name }
        if (kde != null) {
            if (kde is T) {
                return kde
            }
        }
        return null
    }

    inline fun <reified T: IEventKDE> getKDE(): T? {
        val kde = kdes.find { it.ValueType == T::class.java }
        if (kde != null) {
            if (kde is T) {
                return kde
            }
        }
        return null
    }
}
